[Bug]: special orders crossed panic in market-sim fuzz testing

### Problem encountered

vega-market-sim job triggered by a push to develop encountered a panic in the fuzz-testing.

- [node.err](https://jenkins.vega.rocks/job/common/job/vega-market-sim/1143/artifact/network_home/vega-sim-wrmzj_zr/node.err/*view*)
- [node.out](https://jenkins.vega.rocks/job/common/job/vega-market-sim/1143/artifact/network_home/vega-sim-wrmzj_zr/node.out/*view*)
- [replay](https://jenkins.vega.rocks/job/common/job/vega-market-sim/1143/artifact/network_home/vega-sim-wrmzj_zr/vegahome/replay/*view*)

```
panic: re submit special order would cross

goroutine 34908 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x2?, 0x2?, {0x0?, 0x0?, 0xc00b465000?})
	/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:198 +0x65
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00143a1a0, {0xc00e6fc180, 0x2, 0x2})
	/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:264 +0x3ec
go.uber.org/zap.(*Logger).Panic(0x2bb838787f298000?, {0x3a0de9a?, 0x0?}, {0xc00e6fc180, 0x2, 0x2})
	/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:258 +0x59
code.vegaprotocol.io/vega/core/matching.(*OrderBook).ReSubmitSpecialOrders(0xc00508d720, 0xc00e4924e0)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/matching/orderbook.go:796 +0xda5
code.vegaprotocol.io/vega/core/execution/future.(*Market).reSubmitPeggedOrders(0xc0002a4280, {0x40d5a88, 0xc00267d920}, {0xc003595600, 0x2e, 0x721285?})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/future/special_orders.go:110 +0xf7
code.vegaprotocol.io/vega/core/execution/future.(*Market).repriceAllSpecialOrders(0xc0002a4280, {0x40d5a88, 0xc00267d920}, 0x40?, {0xc0099a49d0?, 0xc0099a49d0?, 0x10?})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/future/special_orders.go:149 +0x12e
code.vegaprotocol.io/vega/core/execution/future.(*Market).checkForReferenceMoves(0xc0002a4280, {0x40d5a88, 0xc00267d920}, {0xc0099a49d0, 0x2, 0x2}, 0x0)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/future/reference_price_moves.go:55 +0x2e7
code.vegaprotocol.io/vega/core/execution/future.(*Market).SubmitOrderWithIDGeneratorAndOrderID(0xc0002a4280, {0x40d5a88, 0xc00267d920}, 0x1?, {0xc009509f40, 0x40}, {0x40a24a0?, 0xc00a60d6e0?}, {0xc0097be980, 0x40}, ...)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/future/market.go:1863 +0x6db
code.vegaprotocol.io/vega/core/execution.(*Engine).SubmitOrder(0xc0003f3600, {0x40d5a88, 0xc00267d920}, 0xc00ce5e240, {0xc009509f40, 0x40}, {0x40a24a0, 0xc00a60d6e0}, {0xc0097be980, 0x40})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/engine.go:1057 +0x4af
code.vegaprotocol.io/vega/core/processor.(*BMIProcessor).ProcessBatch(0xc00671f4f8, {0x40d5a88, 0xc00267d920}, 0xc0035f4c80, {0xc009509f40, 0x40}, {0xc0097be840, 0x40}, {0x4118ea8, 0xc00135b440})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/batch_market_instructions_processor.go:187 +0xa96
code.vegaprotocol.io/vega/core/processor.(*App).DeliverBatchMarketInstructions(0xc0014a0380, {0x40d5a88, 0xc00267d920}, {0x40f8270, 0xc00409a980}, {0xc0097be840, 0x40})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/abci.go:1357 +0x167
code.vegaprotocol.io/vega/core/processor.addDeterministicID.func1({0x40d5a88, 0xc00267d920}, {0x40f8270, 0xc00409a980})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/abci.go:489 +0x103
code.vegaprotocol.io/vega/core/processor.(*App).CheckBatchMarketInstructionsW.func1({0x40d5a88, 0xc00267d920}, {0x40f8270, 0xc00409a980})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/abci.go:511 +0x76
code.vegaprotocol.io/vega/core/processor.(*App).SendTransactionResult.func1({0x40d5a88, 0xc00267d920}, {0x40f8270, 0xc00409a980})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/abci.go:541 +0x5c
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).DeliverTx(0xc0012564d0, {{0xc002b3ed80, 0xcaa, 0xcaa}})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/abci/abci.go:140 +0x762
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).DeliverTx(0xc00f79cef8?, {{0xc002b3ed80?, 0x0?, 0x0?}})
	/jenkins/workspace/common/vega-market-sim/extern/vega/cmd/vega/commands/node/app_wrapper.go:37 +0x6c
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).processBlock(0xc0012eca20)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:197 +0x2a3
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).ForwardTime(0xc0012eca20, 0x3b9aca00)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:272 +0x1e5
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime.func1()
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:129 +0x2a
created by code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:128 +0x24a
```

### Observed behaviour

Panics

### Expected behaviour

Should not panic

### Steps to reproduce

```bash
Replay
```


### Software version

9dab1ac2fc6d23a3159f5b25ca33810a258efc18

### Failing test

tests/integration/test_fuzz.py::test_fuzz_run

### Jenkins run

https://jenkins.vega.rocks/job/common/job/vega-market-sim/1143/display/redirect

### Configuration used

```shell
{
  "genesis_time": "2023-05-09T10:00:00.000000000Z",
  "chain_id": "CUSTOM",
  "initial_height": "0",
  "consensus_params": {
    "block": {
      "max_bytes": "22020096",
      "max_gas": "-1",
      "time_iota_ms": "1"
    },
    "evidence": {
      "max_age_num_blocks": "100000",
      "max_age_duration": "172800000000000",
      "max_bytes": "1048576"
    },
    "validator": {
      "pub_key_types": ["ed25519"]
    },
    "version": {}
  },
  "validators": [
    {
      "address": "7DFD3744A773C166B36F170E4F5EF3E998703F4E",
      "pub_key": {
        "type": "tendermint/PubKeyEd25519",
        "value": "2lPO6I0p5uCBZzeoA2+ECJ2+57LtLVMjXpxH5za9TvE="
      },
      "power": "10",
      "name": ""
    }
  ],
  "app_hash": "",
  "app_state": {
    "assets": {
      "VOTE": {
        "name": "VOTE",
        "symbol": "VOTE",
        "decimals": 5,
        "min_lp_stake": "1",
        "source": {
          "builtin_asset": {
            "max_faucet_amount_mint": "100000000000"
          }
        }
      }
    },
    "validators": {
      "6NemqXSfN4zfZvAQeIANflEFqQ98izX9C0KLI2dITSY=": {
        "id": "c2b0ab69fb1bd19862c04df401920ea62d3e0bce41a50148240f42017824d17a",
        "vega_pub_key": "dfcf0fa8d67cfa11b3a511e320f3c47807a5e0735cc06dd9c2bc00f7ccc67117",
        "vega_pub_key_index": 1,
        "ethereum_address": "0x9e5BEEC6E56B28cCbd02864840B0f1e0125e42Ce",
        "tm_pub_key": "6NemqXSfN4zfZvAQeIANflEFqQ98izX9C0KLI2dITSY=",
        "info_url": "",
        "country": "",
        "name": "",
        "avatar_url": ""
      }
    },
    "network": {
      "replay_attack_threshold": 150
    },
    "network_parameters": {
      "blockchains.ethereumConfig": "{\"network_id\": \"3\", \"chain_id\": \"3\", \"collateral_bridge_contract\": { \"address\": \"0xa6F1E140daC13002Dfd9789D6dBA59117c717D7a\" }, \"confirmations\": 50, \"staking_bridge_contract\": { \"address\": \"0xfce2CC92203A266a9C8e67461ae5067c78f67235\", \"deployment_block_height\": 11001702}, \"multisig_control_contract\": {\"address\": \"0xCF6d41235911184fe6F35D47207813bFF3B91601\", \"deployment_block_height\": 12710009 } }",
      "governance.proposal.asset.maxClose": "8760h0m0s",
      "governance.proposal.asset.maxEnact": "8760h0m0s",
      "governance.proposal.asset.minClose": "1s",
      "governance.proposal.asset.minEnact": "1s",
      "governance.proposal.asset.minProposerBalance": "1",
      "governance.proposal.asset.minVoterBalance": "1",
      "governance.proposal.asset.requiredMajority": "0.66",
      "governance.proposal.asset.requiredParticipation": "0.00001",
      "governance.proposal.freeform.maxClose": "8760h0m0s",
      "governance.proposal.freeform.minClose": "1s",
      "governance.proposal.freeform.minProposerBalance": "1",
      "governance.proposal.freeform.minVoterBalance": "1",
      "governance.proposal.freeform.requiredMajority": "0.66",
      "governance.proposal.freeform.requiredParticipation": "0.00001",
      "governance.proposal.market.maxClose": "8760h0m0s",
      "governance.proposal.market.maxEnact": "8760h0m0s",
      "governance.proposal.market.minClose": "1s",
      "governance.proposal.market.minEnact": "2s",
      "governance.proposal.market.minProposerBalance": "1",
      "governance.proposal.market.minVoterBalance": "1",
      "governance.proposal.market.requiredMajority": "0.66",
      "governance.proposal.market.requiredParticipation": "0.00001",
      "governance.proposal.updateAsset.maxClose": "720h",
      "governance.proposal.updateAsset.maxEnact": "720h",
      "governance.proposal.updateAsset.minClose": "1s",
      "governance.proposal.updateAsset.minEnact": "1s",
      "governance.proposal.updateAsset.minProposerBalance": "1",
      "governance.proposal.updateAsset.minVoterBalance": "1",
      "governance.proposal.updateAsset.requiredMajority": "0.66",
      "governance.proposal.updateAsset.requiredParticipation": "0.09",
      "governance.proposal.updateMarket.maxClose": "8760h0m0s",
      "governance.proposal.updateMarket.maxEnact": "8760h0m0s",
      "governance.proposal.updateMarket.minClose": "1s",
      "governance.proposal.updateMarket.minEnact": "1s",
      "governance.proposal.updateMarket.minProposerBalance": "1",
      "governance.proposal.updateMarket.minProposerEquityLikeShare": "0",
      "governance.proposal.updateMarket.minVoterBalance": "1",
      "governance.proposal.updateMarket.requiredMajority": "0.66",
      "governance.proposal.updateMarket.requiredMajorityLP": "0.66",
      "governance.proposal.updateMarket.requiredParticipation": "0.00001",
      "governance.proposal.updateMarket.requiredParticipationLP": "0",
      "governance.proposal.updateNetParam.maxClose": "8760h0m0s",
      "governance.proposal.updateNetParam.maxEnact": "8760h0m0s",
      "governance.proposal.updateNetParam.minClose": "1s",
      "governance.proposal.updateNetParam.minEnact": "1s",
      "governance.proposal.updateNetParam.minProposerBalance": "1",
      "governance.proposal.updateNetParam.minVoterBalance": "1",
      "governance.proposal.updateNetParam.requiredMajority": "0.5",
      "governance.proposal.updateNetParam.requiredParticipation": "0.00001",
      "governance.proposal.referralProgram.maxClose": "8760h0m0s",
      "governance.proposal.referralProgram.maxEnact": "8760h0m0s",
      "governance.proposal.referralProgram.minClose": "1s",
      "governance.proposal.referralProgram.minEnact": "1s",
      "governance.proposal.referralProgram.minProposerBalance": "1",
      "governance.proposal.referralProgram.minVoterBalance": "1",
      "governance.proposal.referralProgram.requiredMajority": "0.5",
      "governance.proposal.referralProgram.requiredParticipation": "0.00001",
      "governance.proposal.VolumeDiscountProgram.maxClose": "8760h0m0s",
      "governance.proposal.VolumeDiscountProgram.maxEnact": "8760h0m0s",
      "governance.proposal.VolumeDiscountProgram.minClose": "1s",
      "governance.proposal.VolumeDiscountProgram.minEnact": "1s",
      "governance.proposal.VolumeDiscountProgram.minProposerBalance": "1",
      "governance.proposal.VolumeDiscountProgram.minVoterBalance": "1",
      "governance.proposal.VolumeDiscountProgram.requiredMajority": "0.5",
      "governance.proposal.VolumeDiscountProgram.requiredParticipation": "0.00001",
      "limits.markets.maxPeggedOrders": "1500",
      "market.auction.maximumDuration": "168h",
      "market.auction.minimumDuration": "1s",
      "market.fee.factors.infrastructureFee": "0.0005",
      "market.fee.factors.makerFee": "0.0002",
      "market.liquidity.bondPenaltyParameter": "0.1",
      "market.liquidity.earlyExitPenalty": "0.1",
      "market.liquidity.maximumLiquidityFeeFactorLevel": "0.03",
      "market.liquidity.minimum.probabilityOfTrading.lpOrders": "1e-6",
      "market.liquidity.probabilityOfTrading.tau.scaling": "10.0",
      "market.liquidity.providersFeeCalculationTimeStep": "1s",
      "market.liquidity.sla.nonPerformanceBondPenaltyMax": "0.5",
      "market.liquidity.sla.nonPerformanceBondPenaltySlope": "2",
      "market.liquidity.stakeToCcyVolume": "1.0",
      "market.liquidity.targetstake.triggering.ratio": "0.25",
      "market.liquidityProvision.minLpStakeQuantumMultiple": "5000",
      "market.liquidityProvision.shapes.maxSize": "100",
      "market.margin.scalingFactors": "{\"search_level\": 1.1, \"initial_margin\": 1.5, \"collateral_release\": 1.7}",
      "market.monitor.price.defaultParameters": "{\"triggers\": [{\"auction_extension\": 300, \"horizon\": 43200, \"probability\": \"0.9999999\"}] }",
      "market.stake.target.scalingFactor": "0.0001",
      "market.stake.target.timeWindow": "1h",
      "market.value.windowLength": "2h0m0s",
      "network.checkpoint.timeElapsedBetweenCheckpoints": "5m",
      "network.floatingPointUpdates.delay": "5m",
      "network.markPriceUpdateMaximumFrequency": "5s",
      "network.transaction.defaultgas": "1",
      "network.transactions.maxgasperblock": "3000000",
      "network.transactions.minBlockCapacity": "32",
      "network.validators.ersatz.multipleOfTendermintValidators": "0",
      "network.validators.ersatz.rewardFactor": "0.85",
      "network.validators.incumbentBonus": "0.05",
      "network.validators.minimumEthereumEventsForNewValidator": "3",
      "network.validators.multisig.numberOfSigners": "13",
      "network.validators.tendermint.number": "13",
      "reward.asset": "VOTE",
      "reward.staking.delegation.competitionLevel": "1.3",
      "reward.staking.delegation.delegatorShare": "0.8",
      "reward.staking.delegation.maxPayoutPerEpoch": "9890000000000000000000",
      "reward.staking.delegation.maxPayoutPerParticipant": "700000000000000000000",
      "reward.staking.delegation.minimumValidatorStake": "3000000000000000000000",
      "reward.staking.delegation.minValidators": "5",
      "reward.staking.delegation.optimalStakeMultiplier": "3.0",
      "reward.staking.delegation.payoutDelay": "0h",
      "reward.staking.delegation.payoutFraction": "1.0",
      "rewards.marketCreationQuantumMultiple": "50000000",
      "snapshot.interval.length": "300",
      "spam.pow.difficulty": "15",
      "spam.pow.hashFunction": "sha3_24_rounds",
      "spam.pow.increaseDifficulty": "0",
      "spam.pow.numberOfPastBlocks": "100",
      "spam.pow.numberOfTxPerBlock": "1000",
      "spam.protection.delegation.min.tokens": "100000000000000000",
      "spam.protection.max.batchSize": "30",
      "spam.protection.max.delegations": "360",
      "spam.protection.max.proposals": "300",
      "spam.protection.max.votes": "300",
      "spam.protection.maxUserTransfersPerEpoch": "20",
      "spam.protection.minimumWithdrawalQuantumMultiple": "10",
      "spam.protection.minMultisigUpdates": "100000000000000000000",
      "spam.protection.proposal.min.tokens": "2000000000000000000000",
      "spam.protection.voting.min.tokens": "1000000000000000000",
      "transfer.fee.factor": "0.001",
      "transfer.minTransferQuantumMultiple": "100",
      "validator.performance.scaling.factor": "0",
      "validators.delegation.minAmount": "1",
      "validators.epoch.length": "5m",
      "validators.vote.required": "0.67",
      "referralProgram.maxReferralTiers": "10",
      "referralProgram.maxReferralRewardFactor": "1",
      "referralProgram.maxReferralDiscountFactor": "1",
      "referralProgram.maxPartyNotionalVolumeByQuantumPerEpoch": "10000",
      "referralProgram.minStakedVegaTokens": "0",
      "referralProgram.maxReferralRewardProportion": "1"
    },
    "network_limits": {
      "propose_market_enabled": true,
      "propose_asset_enabled": true,
      "bootstrap_block_count": 0
    },
    "checkpoint": {
      "load_hash": ""
    }
  }
}
```


### Relevant log output

```shell
panic: re submit special order would cross

goroutine 34908 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x2?, 0x2?, {0x0?, 0x0?, 0xc00b465000?})
	/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:198 +0x65
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00143a1a0, {0xc00e6fc180, 0x2, 0x2})
	/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:264 +0x3ec
go.uber.org/zap.(*Logger).Panic(0x2bb838787f298000?, {0x3a0de9a?, 0x0?}, {0xc00e6fc180, 0x2, 0x2})
	/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:258 +0x59
code.vegaprotocol.io/vega/core/matching.(*OrderBook).ReSubmitSpecialOrders(0xc00508d720, 0xc00e4924e0)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/matching/orderbook.go:796 +0xda5
code.vegaprotocol.io/vega/core/execution/future.(*Market).reSubmitPeggedOrders(0xc0002a4280, {0x40d5a88, 0xc00267d920}, {0xc003595600, 0x2e, 0x721285?})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/future/special_orders.go:110 +0xf7
code.vegaprotocol.io/vega/core/execution/future.(*Market).repriceAllSpecialOrders(0xc0002a4280, {0x40d5a88, 0xc00267d920}, 0x40?, {0xc0099a49d0?, 0xc0099a49d0?, 0x10?})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/future/special_orders.go:149 +0x12e
code.vegaprotocol.io/vega/core/execution/future.(*Market).checkForReferenceMoves(0xc0002a4280, {0x40d5a88, 0xc00267d920}, {0xc0099a49d0, 0x2, 0x2}, 0x0)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/future/reference_price_moves.go:55 +0x2e7
code.vegaprotocol.io/vega/core/execution/future.(*Market).SubmitOrderWithIDGeneratorAndOrderID(0xc0002a4280, {0x40d5a88, 0xc00267d920}, 0x1?, {0xc009509f40, 0x40}, {0x40a24a0?, 0xc00a60d6e0?}, {0xc0097be980, 0x40}, ...)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/future/market.go:1863 +0x6db
code.vegaprotocol.io/vega/core/execution.(*Engine).SubmitOrder(0xc0003f3600, {0x40d5a88, 0xc00267d920}, 0xc00ce5e240, {0xc009509f40, 0x40}, {0x40a24a0, 0xc00a60d6e0}, {0xc0097be980, 0x40})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/execution/engine.go:1057 +0x4af
code.vegaprotocol.io/vega/core/processor.(*BMIProcessor).ProcessBatch(0xc00671f4f8, {0x40d5a88, 0xc00267d920}, 0xc0035f4c80, {0xc009509f40, 0x40}, {0xc0097be840, 0x40}, {0x4118ea8, 0xc00135b440})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/batch_market_instructions_processor.go:187 +0xa96
code.vegaprotocol.io/vega/core/processor.(*App).DeliverBatchMarketInstructions(0xc0014a0380, {0x40d5a88, 0xc00267d920}, {0x40f8270, 0xc00409a980}, {0xc0097be840, 0x40})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/abci.go:1357 +0x167
code.vegaprotocol.io/vega/core/processor.addDeterministicID.func1({0x40d5a88, 0xc00267d920}, {0x40f8270, 0xc00409a980})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/abci.go:489 +0x103
code.vegaprotocol.io/vega/core/processor.(*App).CheckBatchMarketInstructionsW.func1({0x40d5a88, 0xc00267d920}, {0x40f8270, 0xc00409a980})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/abci.go:511 +0x76
code.vegaprotocol.io/vega/core/processor.(*App).SendTransactionResult.func1({0x40d5a88, 0xc00267d920}, {0x40f8270, 0xc00409a980})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/processor/abci.go:541 +0x5c
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).DeliverTx(0xc0012564d0, {{0xc002b3ed80, 0xcaa, 0xcaa}})
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/abci/abci.go:140 +0x762
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).DeliverTx(0xc00f79cef8?, {{0xc002b3ed80?, 0x0?, 0x0?}})
	/jenkins/workspace/common/vega-market-sim/extern/vega/cmd/vega/commands/node/app_wrapper.go:37 +0x6c
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).processBlock(0xc0012eca20)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:197 +0x2a3
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).ForwardTime(0xc0012eca20, 0x3b9aca00)
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:272 +0x1e5
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime.func1()
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:129 +0x2a
created by code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime
	/jenkins/workspace/common/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:128 +0x24a
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -274,6 +274,7 @@
 - [8737](https://github.com/vegaprotocol/vega/issues/8737) - Fix `GetStopOrder` errors when order has a state change.
 - [8727](https://github.com/vegaprotocol/vega/issues/8727) - Clear parent market on checkpoint restore if the parent market was already succeeded.
 - [8835](https://github.com/vegaprotocol/vega/issues/8835) - Spot snapshot fixes
+- [9651](https://github.com/vegaprotocol/vega/issues/9651) - Park pegged orders if they lose their peg
 
 ## 0.72.0
 

--- a/core/execution/future/event_generation_test.go
+++ b/core/execution/future/event_generation_test.go
@@ -426,13 +426,10 @@ func TestEvents_PeggedOrderNotAbleToRepriceDueToMargin(t *testing.T) {
 	require.NoError(t, err)
 
 	// Check we have the right amount of events
-	// assert.Equal(t, uint64(6), tm.orderEventCount)
-	assert.Equal(t, uint64(5), tm.orderEventCount)
-	// assert.Equal(t, int64(2), tm.market.GetOrdersOnBookCount())
+	assert.Equal(t, uint64(6), tm.orderEventCount)
 	assert.Equal(t, int64(3), tm.market.GetOrdersOnBookCount())
 
 	processEvents(t, tm, mdb)
-	// assert.Equal(t, int64(2), mdb.GetOrderCount(tm.market.GetID()))
 	assert.Equal(t, int64(3), mdb.GetOrderCount(tm.market.GetID()))
 	assert.Equal(t, 1, tm.market.GetPeggedOrderCount())
 	assert.Equal(t, 0, tm.market.GetParkedOrderCount())

--- a/core/execution/future/market_test.go
+++ b/core/execution/future/market_test.go
@@ -4133,11 +4133,11 @@ func TestOrderBook_ExpiredOrderTriggersReprice(t *testing.T) {
 
 		expects := map[string]types.OrderStatus{
 			o1.ID: types.OrderStatusExpired,
-			o2.ID: types.OrderStatusActive,
+			o2.ID: types.OrderStatusParked,
 		}
 
 		for id, v := range found {
-			require.Equal(t, v.Status, expects[id])
+			require.Equal(t, expects[id], v.Status)
 		}
 	})
 }

--- a/core/execution/future/order_test.go
+++ b/core/execution/future/order_test.go
@@ -1972,7 +1972,8 @@ func testPeggedOrderExpiring2(t *testing.T) {
 	tm.now = tm.now.Add(time.Second)
 	tm.market.OnTick(ctx, tm.now)
 
-	assert.Equal(t, 1, tm.market.GetParkedOrderCount())
+	// the limit order has expired so the order pegged to it has to be parked!
+	assert.Equal(t, 2, tm.market.GetParkedOrderCount())
 	assert.Equal(t, 2, tm.market.GetPeggedOrderCount())
 
 	// Move the time forward

--- a/core/execution/future/reference_price_moves.go
+++ b/core/execution/future/reference_price_moves.go
@@ -26,9 +26,6 @@ func (m *Market) checkForReferenceMoves(
 		return
 	}
 
-	// will be set to non-nil if a peg is missing
-	_, _, err := m.getBestStaticPricesDecimal()
-
 	newBestBid, _ := m.getBestStaticBidPrice()
 	newBestAsk, _ := m.getBestStaticAskPrice()
 	newMidBuy, _ := m.getStaticMidPrice(types.SideBuy)
@@ -51,11 +48,7 @@ func (m *Market) checkForReferenceMoves(
 	}
 
 	// now we can start all special order repricing...
-	if err == nil {
-		orderUpdates = m.repriceAllSpecialOrders(ctx, changes, orderUpdates)
-	} else {
-		orderUpdates = nil
-	}
+	orderUpdates = m.repriceAllSpecialOrders(ctx, changes, orderUpdates)
 
 	// Update the last price values
 	// no need to clone the prices, they're not used in calculations anywhere in this function

--- a/core/execution/spot/reference_price_moves.go
+++ b/core/execution/spot/reference_price_moves.go
@@ -24,9 +24,6 @@ func (m *Market) checkForReferenceMoves(ctx context.Context, forceUpdate bool) {
 		return
 	}
 
-	// will be set to non-nil if a peg is missing
-	_, _, err := m.getBestStaticPricesDecimal()
-
 	newBestBid, _ := m.getBestStaticBidPrice()
 	newBestAsk, _ := m.getBestStaticAskPrice()
 	newMidBuy, _ := m.getStaticMidPrice(types.SideBuy)
@@ -48,13 +45,7 @@ func (m *Market) checkForReferenceMoves(ctx context.Context, forceUpdate bool) {
 		changes = common.PriceMoveAll
 	}
 
-	// now we can start all special order repricing...
-	if err == nil {
-		m.repriceAllSpecialOrders(ctx, changes)
-	} else {
-		// we won't be able to reprice here
-		m.parkAllPeggedOrders(ctx)
-	}
+	m.repriceAllSpecialOrders(ctx, changes)
 
 	// Update the last price values
 	// no need to clone the prices, they're not used in calculations anywhere in this function

--- a/core/integration/features/auctions/5460-market-order-cannot-trade-due-to-auction.feature
+++ b/core/integration/features/auctions/5460-market-order-cannot-trade-due-to-auction.feature
@@ -2,14 +2,14 @@ Feature: Test for issue 5460
 
   Background:
     Given the following network parameters are set:
-      | name                                          | value |
-      | network.floatingPointUpdates.delay            | 10s   |
-      | market.auction.minimumDuration                | 1     |
-      | limits.markets.maxPeggedOrders                | 2     |
+      | name                               | value |
+      | network.floatingPointUpdates.delay | 10s   |
+      | market.auction.minimumDuration     | 1     |
+      | limits.markets.maxPeggedOrders     | 2     |
     And the liquidity monitoring parameters:
-      | name               | triggering ratio | time window | scaling factor |
-      | lqm-params         | 1.00             | 24h         | 1              |  
-    
+      | name       | triggering ratio | time window | scaling factor |
+      | lqm-params | 1.00             | 24h         | 1              |
+
     And the following assets are registered:
       | id  | decimal places |
       | ETH | 5              |
@@ -45,9 +45,9 @@ Feature: Test for issue 5460
       | lp1 | party0 | ETH/DEC21 | 200000000         | 0.001 | submission |
       | lp1 | party0 | ETH/DEC21 | 200000000         | 0.001 | submission |
     And the parties place the following pegged iceberg orders:
-      | party  | market id | peak size | minimum visible size | side | pegged reference | volume   | offset |
-      | party0 | ETH/DEC21 | 1000      | 1                    | buy  | MID              | 200000   | 205    |
-      | party0 | ETH/DEC21 | 1000      | 1                    | sell | MID              | 300000   | 205    |
+      | party  | market id | peak size | minimum visible size | side | pegged reference | volume | offset |
+      | party0 | ETH/DEC21 | 1000      | 1                    | buy  | MID              | 200000 | 205    |
+      | party0 | ETH/DEC21 | 1000      | 1                    | sell | MID              | 300000 | 205    |
     And the parties place the following orders:
       | party    | market id | side | volume | price | resulting trades | type       | tif     |
       | party_a1 | ETH/DEC21 | buy  | 100000 | 30000 | 0                | TYPE_LIMIT | TIF_GTC |
@@ -97,7 +97,6 @@ Feature: Test for issue 5460
       | buy  | 29977 | 100000 |
       | buy  | 29967 | 100000 |
       | buy  | 29957 | 100000 |
-      | buy  | 29795 | 1000   |
       | sell | 30002 | 0      |
       | sell | 30204 | 0      |
     When the network moves ahead "1" blocks
@@ -113,32 +112,32 @@ Feature: Test for issue 5460
     Then the market state should be "STATE_ACTIVE" for the market "ETH/DEC21"
 
     And the order book should have the following volumes for market "ETH/DEC21":
-      | side | price | volume    |
-      | buy  | 29998 | 100000    |
-      | buy  | 29987 | 100000    |
-      | buy  | 29977 | 100000    |
-      | buy  | 29967 | 100000    |
-      | buy  | 29957 | 100000    |
-      | buy  | 29795 | 1000      |
-      | sell | 30002 | 100000    |
-      | sell | 30205 | 1000      |
+      | side | price | volume |
+      | buy  | 29998 | 100000 |
+      | buy  | 29987 | 100000 |
+      | buy  | 29977 | 100000 |
+      | buy  | 29967 | 100000 |
+      | buy  | 29957 | 100000 |
+      | buy  | 29795 | 1000   |
+      | sell | 30002 | 100000 |
+      | sell | 30205 | 1000   |
 
     And the parties place the following orders:
       | party   | market id | side | volume | price | resulting trades | type       | tif     |
       | party_r | ETH/DEC21 | buy  | 100000 | 29700 | 0                | TYPE_LIMIT | TIF_GTC |
 
     And the order book should have the following volumes for market "ETH/DEC21":
-      | side | price  | volume    |
-      | buy  | 29998  | 100000    |
-      | buy  | 29987  | 100000    |
-      | buy  | 29977  | 100000    |
-      | buy  | 29967  | 100000    |
-      | buy  | 29957  | 100000    |
-      | buy  | 29795  | 1000      |
-      | buy  | 400000 | 0         |
-      | buy  | 29700  | 100000    |
-      | sell | 30002  | 100000    |
-      | sell | 30205  | 1000      |
+      | side | price  | volume |
+      | buy  | 29998  | 100000 |
+      | buy  | 29987  | 100000 |
+      | buy  | 29977  | 100000 |
+      | buy  | 29967  | 100000 |
+      | buy  | 29957  | 100000 |
+      | buy  | 29795  | 1000   |
+      | buy  | 400000 | 0      |
+      | buy  | 29700  | 100000 |
+      | sell | 30002  | 100000 |
+      | sell | 30205  | 1000   |
 
     And the market data for the market "ETH/DEC21" should be:
       | trading mode            | auction trigger             | target stake | supplied stake | open interest |

--- a/core/integration/features/verified/0003-MTMK-mark_to_market_settlement_using_insurance_pool.feature
+++ b/core/integration/features/verified/0003-MTMK-mark_to_market_settlement_using_insurance_pool.feature
@@ -26,9 +26,9 @@ Feature: Test mark to market settlement with insurance pool
       | lp1 | lpprov | ETH/DEC19 | 90000             | 0.1 | submission |
       | lp1 | lpprov | ETH/DEC19 | 90000             | 0.1 | submission |
     And the parties place the following pegged iceberg orders:
-      | party  | market id | peak size | minimum visible size | side | pegged reference | volume | offset |
-      | lpprov | ETH/DEC19 | 90        | 1                    | buy  | BID              | 90     | 10     |
-      | lpprov | ETH/DEC19 | 15        | 1                    | sell | ASK              | 15     | 10     |
+      | party  | market id | peak size | minimum visible size | side | pegged reference | volume | offset | reference |
+      | lpprov | ETH/DEC19 | 90        | 1                    | buy  | BID              | 90     | 10     | lp-buy    |
+      | lpprov | ETH/DEC19 | 15        | 1                    | sell | ASK              | 15     | 10     | lp-sell   |
 
     # place auxiliary orders so we always have best bid and best offer as to not trigger the liquidity auction
     When the parties place the following orders:
@@ -73,8 +73,18 @@ Feature: Test mark to market settlement with insurance pool
       | party2 | ETH   | ETH/DEC19 | 13598  | 1302    |
       | party3 | ETH   | ETH/DEC19 | 7920   | 1480    |
 
+    #party1 is closed out and traded with sell order ref-2, so there is no best ask to peg
+    And the orders should have the following status:
+      | party  | reference | status        |
+      | aux    | ref-2     | STATUS_FILLED |
+      | lpprov | lp-buy    | STATUS_ACTIVE |
+      | lpprov | lp-sell   | STATUS_PARKED |
+
+    And the trading mode should be "TRADING_MODE_CONTINUOUS" for the market "ETH/DEC19"
+
     And the cumulated balance for all accounts should be worth "155122"
     And the settlement account should have a balance of "0" for the market "ETH/DEC19"
     And the insurance pool balance should be "9999" for the market "ETH/DEC19"
+
 
 

--- a/core/integration/features/verified/0003-MTMK-mark_to_market_settlement_using_insurance_pool.feature
+++ b/core/integration/features/verified/0003-MTMK-mark_to_market_settlement_using_insurance_pool.feature
@@ -26,9 +26,9 @@ Feature: Test mark to market settlement with insurance pool
       | lp1 | lpprov | ETH/DEC19 | 90000             | 0.1 | submission |
       | lp1 | lpprov | ETH/DEC19 | 90000             | 0.1 | submission |
     And the parties place the following pegged iceberg orders:
-      | party  | market id | peak size | minimum visible size | side | pegged reference | volume     | offset |
-      | lpprov | ETH/DEC19 | 90 | 1 | buy  | BID | 90 | 10 |
-      | lpprov | ETH/DEC19 | 15 | 1 | sell | ASK | 15 | 10 |
+      | party  | market id | peak size | minimum visible size | side | pegged reference | volume | offset |
+      | lpprov | ETH/DEC19 | 90        | 1                    | buy  | BID              | 90     | 10     |
+      | lpprov | ETH/DEC19 | 15        | 1                    | sell | ASK              | 15     | 10     |
 
     # place auxiliary orders so we always have best bid and best offer as to not trigger the liquidity auction
     When the parties place the following orders:
@@ -41,11 +41,11 @@ Feature: Test mark to market settlement with insurance pool
     And the trading mode should be "TRADING_MODE_CONTINUOUS" for the market "ETH/DEC19"
     And the market data for the market "ETH/DEC19" should be:
       | mark price | trading mode            | target stake | supplied stake | open interest |
-      | 1000 | TRADING_MODE_CONTINUOUS | 1100 | 88120 | 1 |
+      | 1000       | TRADING_MODE_CONTINUOUS | 1100         | 88120          | 1             |
     Then the parties should have the following account balances:
       | party  | asset | market id | margin | general | bond  |
       | lpprov | ETH   | ETH/DEC19 | 11880  | 0       | 88120 |
-     
+
     And the settlement account should have a balance of "0" for the market "ETH/DEC19"
     When the parties place the following orders with ticks:
       | party  | market id | side | volume | price | resulting trades | type       | tif     | reference |
@@ -54,7 +54,7 @@ Feature: Test mark to market settlement with insurance pool
     Then the parties should have the following account balances:
       | party  | asset | market id | margin | general |
       | party1 | ETH   | ETH/DEC19 | 1320   | 3802    |
-      | party2 | ETH | ETH/DEC19 | 132 | 9768 |
+      | party2 | ETH   | ETH/DEC19 | 132    | 9768    |
 
     And the settlement account should have a balance of "0" for the market "ETH/DEC19"
     When the parties place the following orders with ticks:
@@ -62,16 +62,16 @@ Feature: Test mark to market settlement with insurance pool
       | party2 | ETH/DEC19 | buy  | 1      | 6000  | 0                | TYPE_LIMIT | TIF_GTC | ref-1     |
     Then the parties should have the following account balances:
       | party  | asset | market id | margin | general |
-      | party2 | ETH | ETH/DEC19 | 265 | 9635 |
+      | party2 | ETH   | ETH/DEC19 | 265    | 9635    |
 
     When the parties place the following orders with ticks:
       | party  | market id | side | volume | price | resulting trades | type       | tif     | reference |
-      | party3 | ETH/DEC19 | sell | 1 | 6000 | 1 | TYPE_LIMIT | TIF_GTC | ref-3 |
+      | party3 | ETH/DEC19 | sell | 1      | 6000  | 1                | TYPE_LIMIT | TIF_GTC | ref-3     |
     Then the parties should have the following account balances:
       | party  | asset | market id | margin | general |
       | party1 | ETH   | ETH/DEC19 | 0      | 0       |
-      | party2 | ETH | ETH/DEC19 | 13598 | 1302 |
-      | party3 | ETH | ETH/DEC19 | 721   | 8679 |
+      | party2 | ETH   | ETH/DEC19 | 13598  | 1302    |
+      | party3 | ETH   | ETH/DEC19 | 7920   | 1480    |
 
     And the cumulated balance for all accounts should be worth "155122"
     And the settlement account should have a balance of "0" for the market "ETH/DEC19"

--- a/core/integration/features/verified/0044-LIME_lp_at_termination.feature
+++ b/core/integration/features/verified/0044-LIME_lp_at_termination.feature
@@ -28,23 +28,23 @@ Feature: Test LP bond account when market is terminated
       | horizon | probability | auction extension |
       | 360000  | 0.99        | 3                 |
     And the following network parameters are set:
-      | name                                                  | value |
-      | market.value.windowLength                             | 60s   |
-      | network.markPriceUpdateMaximumFrequency               | 0s    |
-      | limits.markets.maxPeggedOrders                        | 6     |
-      | market.auction.minimumDuration                        | 1     |
-      | market.fee.factors.infrastructureFee                  | 0.001 |
-      | market.fee.factors.makerFee                           | 0.004 |
+      | name                                                | value |
+      | market.value.windowLength                           | 60s   |
+      | network.markPriceUpdateMaximumFrequency             | 0s    |
+      | limits.markets.maxPeggedOrders                      | 6     |
+      | market.auction.minimumDuration                      | 1     |
+      | market.fee.factors.infrastructureFee                | 0.001 |
+      | market.fee.factors.makerFee                         | 0.004 |
       | market.liquidity.bondPenaltyParameter               | 0.2   |
-      | validators.epoch.length                               | 5s    |
+      | validators.epoch.length                             | 5s    |
       | market.liquidity.stakeToCcyVolume                   | 1     |
-      | market.liquidity.successorLaunchWindowLength          | 1h    |
+      | market.liquidity.successorLaunchWindowLength        | 1h    |
       | market.liquidity.sla.nonPerformanceBondPenaltySlope | 0.19  |
       | market.liquidity.sla.nonPerformanceBondPenaltyMax   | 1     |
-      | validators.epoch.length                               | 2s    |
+      | validators.epoch.length                             | 2s    |
     Given the liquidity monitoring parameters:
-      | name               | triggering ratio | time window | scaling factor |
-      | lqm-params         | 0.5              | 20s         | 1.0            |
+      | name       | triggering ratio | time window | scaling factor |
+      | lqm-params | 0.5              | 20s         | 1.0            |
 
     And the liquidity sla params named "SLA":
       | price range | commitment min time fraction | performance hysteresis epochs | sla competition factor |
@@ -53,7 +53,7 @@ Feature: Test LP bond account when market is terminated
       | id        | quote name | asset | liquidity monitoring | risk model            | margin calculator   | auction duration | fees          | price monitoring | data source config | linear slippage factor | quadratic slippage factor | sla params |
       | ETH/MAR22 | USD        | USD   | lqm-params           | log-normal-risk-model | margin-calculator-1 | 2                | fees-config-1 | price-monitoring | ethDec19Oracle     | 1e1                    | 1e0                       | SLA        |
     And the following network parameters are set:
-      | name                                               | value |
+      | name                                             | value |
       | market.liquidity.providersFeeCalculationTimeStep | 2s    |
 
     Given the average block duration is "2"
@@ -108,7 +108,7 @@ Feature: Test LP bond account when market is terminated
       | party | asset | market id | margin | general | bond |
       | lp1   | USD   | ETH/MAR22 | 26677  | 129323  | 3000 |
       | lp2   | USD   | ETH/MAR22 | 10671  | 88329   | 1000 |
-    #intial margin lp2: 2*1000*3.5569036*1.5=10670
+    #initial margin lp2: 2*1000*3.5569036*1.5=10670
 
     Then the parties should have the following profit and loss:
       | party  | volume | unrealised pnl | realised pnl |
@@ -141,7 +141,7 @@ Feature: Test LP bond account when market is terminated
     And the parties should have the following account balances:
       | party | asset | market id | margin | general | bond |
       | lp1   | USD   | ETH/MAR22 | 155878 | 145     | 2430 |
-      | lp2   | USD   | ETH/MAR22 | 0      | 99000   | 1000 |
+      | lp2   | USD   | ETH/MAR22 | 0      | 99000   | 810  |
 
     When the oracles broadcast data signed with "0xCAFECAFE19":
       | name               | value |
@@ -150,11 +150,11 @@ Feature: Test LP bond account when market is terminated
     Then the market state should be "STATE_TRADING_TERMINATED" for the market "ETH/MAR22"
     And the market data for the market "ETH/MAR22" should be:
       | mark price | trading mode            | horizon | min bound | max bound | target stake | supplied stake | open interest |
-      | 1120       | TRADING_MODE_NO_TRADING | 360000  | 831       | 1440      | 63739        | 3430           | 16            |
+      | 1120       | TRADING_MODE_NO_TRADING | 360000  | 831       | 1440      | 63739        | 3240           | 16            |
 
     Then the oracles broadcast data signed with "0xCAFECAFE19":
       | name             | value |
-      | prices.ETH.value | 1600 |
+      | prices.ETH.value | 1600  |
     And the network moves ahead "3" blocks
 
     Then the parties should have the following profit and loss:
@@ -167,7 +167,7 @@ Feature: Test LP bond account when market is terminated
     And the parties should have the following account balances:
       | party | asset | market id | margin | general | bond |
       | lp1   | USD   | ETH/MAR22 | 0      | 155592  | 0    |
-      | lp2   | USD   | ETH/MAR22 | 0      | 99810   | 0    |
+      | lp2   | USD   | ETH/MAR22 | 0      | 99657   | 0    |
 
-    And the insurance pool balance should be "1255" for the market "ETH/MAR22"
+    And the insurance pool balance should be "1408" for the market "ETH/MAR22"
 

--- a/core/matching/orderbook.go
+++ b/core/matching/orderbook.go
@@ -789,18 +789,12 @@ func (b *OrderBook) ReSubmitSpecialOrders(order *types.Order) {
 	switch order.Side {
 	case types.SideBuy:
 		price, err := b.GetBestAskPrice()
-		if err != nil {
-			b.log.Panic("tried to re submit special orders in an empty book", logging.Order(order))
-		}
-		if price.LTE(order.Price) {
+		if err == nil && price.LTE(order.Price) {
 			b.log.Panic("re submit special order would cross", logging.Order(order), logging.BigUint("best-ask", price))
 		}
 	case types.SideSell:
 		price, err := b.GetBestBidPrice()
-		if err != nil {
-			b.log.Panic("tried to re submit special orders in an empty book", logging.Order(order))
-		}
-		if price.GTE(order.Price) {
+		if err == nil && price.GTE(order.Price) {
 			b.log.Panic("re submit special order would cross", logging.Order(order), logging.BigUint("best-bid", price))
 		}
 	default:

